### ISSUE TITLE
How to use a Maptiler map - .NET MAUI

### ISSUE THREAD

**Issue by ndegheselle** [Mapsui/Mapsui#1797]
I am using [this sample](https://github.com/Mapsui/Mapsui/blob/master/Samples/Mapsui.Samples.Common/Maps/DataFormats/MapTilerSample.cs) in .NET MAUI, it is working fine but now I want to use my own maptiler files with the following files :

```
./12
    2120
         1444.png
         1445.png
    2121
         1444.png
         1445.png
...
./18
    ./135722
       ...
```
I haven't found anywhere to what the folders number and files number relate to (aside that it is coordinates), should it look like this to be used with Mapsui ?

From what I have read on [tiling scheme](https://documentation.maptiler.com/hc/en-us/articles/4405435316369-Tiling-scheme-in-Folder-output) my files are using OpenStreetMap scheme, I also have different zoom so I modified `GetTileSchema` :
```
    public ITileSchema GetTileSchema()
    {
        var schema = new GlobalSphericalMercator(YAxis.OSM);
        schema.Resolutions.Clear();
        schema.Resolutions[12] = new Resolution(12, 156543.033900000);
        schema.Resolutions[13] = new Resolution(13, 78271.516950000);
        return schema;
    }
```
Problem is the map stay white.

Is it because of the Resolution ? I don't really understand where the `156543.033900000` is coming from.
Or is it because of coordinate ? I do have the bounds of my map but I don't know how to set them on the layer. 

Thanks you kindly for you help !

**Comment by charlenni** [comment]
In your link to the tiling schema, the folder structure is well described: The folder path is {z}/{x}/{y}.{ext}, where {ext} is the extension of tile format (png, jpg, webp, …). {z} is the zoom factor, {x} and {y} the tile position. In your example, you show two root folders, one for zoom level 12, one for zoom level 18. Zoom level 12 contains 2 columns (x = 2120 and 2121) and in these columns you have 2 rows (y = 1444 and 1445). If you want to see, which tiles are in your file, you could use https://mt1.google.com/vt/lyrs=y&x=2120&y=1444&z=12.

I would look for one of the samples that use a MBTiles file, like [MBTilesSample](https://github.com/Mapsui/Mapsui/blob/master/Samples/Mapsui.Samples.Common/Maps/DataFormats/MbTilesSample.cs).

**Comment by ndegheselle** [comment]
@charlenni 

> The folder path is {z}/{x}/{y}.{ext}, where {ext} is the extension of tile format (png, jpg, webp, …). {z} is the zoom factor, {x} and {y} the tile position. 

I was more curious about which coordinate system x and y belong to, how do you convert {x=2120, y=1444} in longitude / latitude (tried to find out more about mt1.google didn't find much).

> I would look for one of the samples that use a MBTiles file, like [MBTilesSample](https://github.com/Mapsui/Mapsui/blob/master/Samples/Mapsui.Samples.Common/Maps/DataFormats/MbTilesSample.cs).

I don't see how MBTiles helps me, I am already using MapTilerSample.cs, is the mbtile format better ?

**Comment by ndegheselle** [comment]
Thanks for the links ! Yes I have exactly the same setup than the MBTilesSample with a folder containing all the sub folders and files.

**Comment by pauldendulk** [comment]
Do you get any exceptions from the Mapsui logger? https://mapsui.com/documentation/logging.html

**Comment by charlenni** [comment]
@pauldendulk It seems, that the TO has not a MBTiles file, but the native PNG files in a folder structure. Do we have a provider for such a scenario?

**Comment by pauldendulk** [comment]
@charlenni Yes, this is shown in the [MapTiler](https://github.com/Mapsui/Mapsui/blob/master/Samples/Mapsui.Samples.Common/Maps/DataFormats/MapTilerSample.cs) sample. MapTiler is a tool I used a long time ago to generate a file pyramid on disk.

**Comment by ndegheselle** [comment]
@pauldendulk 
I have an exception from BruTile "The tile was not found at it's expected location", 
Inside `GetTileAsync(TileInfo tileInfo)` I have `tileInfo.Index` with theses values : `{Col=0, Level=13, Row=1}` so I guess the problem is that I don't tell it which tile it should start on, or the bounds of my map (I only have a specific section of the map contrary to the sample). I have tried to set the `Map.Limiter.PanLimits` to the coordinate of the map but I get the same error.

**Comment by charlenni** [comment]
For zoom level 13 you have 8192 different rows and columns. If you want to get col=0 and row=1, then you get an Arktis tile. Is this what you want? Did this tile exists in your data?

**Comment by pauldendulk** [comment]
> For zoom level 13 you have 8192 different rows and columns. If you want to get col=0 and row=1, then you get an Arktis tile. Is this what you want? Did this tile exists in your data?

I think you meant to type 'arctic' and such a tile would be fully white, that could also be a problem.

You also need to know that the tiles at a high detail level are only shown if you zoom to that detail level. If they would be shown above that level this could result in a huge number of tiles.

**Comment by ndegheselle** [comment]
Okay I finnaly got it working, thank you both for the hints !

I added a openstreetmap layer `map.Layers.Add(Mapsui.Tiling.OpenStreetMap.CreateTileLayer());` before my tile instead of an arctic, like so I can search for my Tiles on the map while moving / zooming,

I also ended up removing the resolution part from the MapTile sample (still confused what it is doing in the original sample) : 
```C#
    public ITileSchema GetTileSchema()
    {
        var schema = new GlobalSphericalMercator(YAxis.OSM);
        schema.Resolutions.Clear();
        // Calculated automatically
        // schema.Resolutions[12] = new Resolution(12, 156543.033900000);
        // schema.Resolutions[13] = new Resolution(13, 78271.516950000);
        return schema;
    }
```
The whole class if anyone is interested :
```C#
using BruTile;
using BruTile.Cache;
using BruTile.FileSystem;
using BruTile.Predefined;
using Project.Maui.Lib.Contexte;
using Project.Maui.Lib.Contexte.Data;
using Mapsui;
using Mapsui.Layers;
using Mapsui.Logging;
using Mapsui.Tiling.Layers;
using Mapsui.UI;
using Mapsui.UI.Maui;
using Mapsui.Utilities;
using Map = Mapsui.Map;

namespace Project.Maui.Views.Ville;

public class MapTilerTileSource : ITileSource
{
    public string TilesPath { get; set; }

    public ITileSchema Schema { get; }
    public string Name { get; }
    public Attribution Attribution { get; } = new Attribution();
    public ITileProvider Provider { get; }

    public MapTilerTileSource(string path)
    {
        TilesPath = path;
        Schema = GetTileSchema();
        Provider = GetTileProvider();
        Name = "MapTiler";
    }

    public async Task<byte[]> GetTileAsync(TileInfo tileInfo)
    {
        return await Provider.GetTileAsync(tileInfo);
    }

    public ITileProvider GetTileProvider()
    {
        return new FileTileProvider(new FileCache(TilesPath, "png"));
    }

    public ITileSchema GetTileSchema()
    {
        var schema = new GlobalSphericalMercator(YAxis.OSM);
        schema.Resolutions.Clear();
        // Calculated automatically
        // schema.Resolutions[12] = new Resolution(12, 156543.033900000);
        // schema.Resolutions[13] = new Resolution(13, 78271.516950000);
        return schema;
    }
}

public partial class Carte : ContentView
{
    public Carte()
	{
        InitializeComponent();

        mapView.ViewportInitialized += MapView_ViewportInitialized;
        Mapsui.Logging.Logger.LogDelegate += LogMapsui;
    }

    private void LogMapsui(LogLevel pLog, string pMessage, Exception pEx)
    {
        Console.WriteLine(pMessage);
    }

    private void MapView_ViewportInitialized(object sender, EventArgs e)
    {
        var map = new Map();

        map.Layers.Add(Mapsui.Tiling.OpenStreetMap.CreateTileLayer());
        map.Layers.Add(CreateLayer());

        mapView.Map = map;
        mapView.RotationLock = true;

        var lMin = new Position(46.74620000, 6.43619988).ToMapsui();
        var lMax = new Position(46.71709961, 6.38600000).ToMapsui();
        var lBounding = new MRect(lMin.X, lMin.Y, lMax.X, lMax.Y);
        
        mapView.Navigator.NavigateTo(lBounding, ScaleMethod.Fit);
    }

    private ILayer CreateLayer()
    {
        return new TileLayer(new MapTilerTileSource("Path"));
    }
}
```
